docs/design/eth-subscribe.md

```diff
@@ -0,0 +1,281 @@
+const wss = new WebSocketServer({ server });
+
+wss.on('connection', function (ws) {
+  ws.on('message', function (msg) {
+    const msgString = msg.toString('base64');
+    const request = JSON.parse(Buffer.from(msgString, 'base64').toString('ascii'));
+    const { method, params } = request;
+
+    if (method === 'eth_subscribe') {
+      const event = params[0];
+
+      if (event === 'log') {
+        const filter = params[1];
+        // Cache the `event`, `filter`, `SUBSCRIPTION_ID` and the current block number values
+
+        // This establishes the subscription
+        ws.send(
+          JSON.stringify({
+            id: request.id,
+            result: SUBSCRIPTION_ID,
+            jsonrpc: '2.0',
+          })
+        );
+      }
+
+      // Handle other types of events: newPendingTransactions, newHeads, syncing
+    }
+
+    // The WebSocketProvider emits an event to `eth_chainId` right after establishing a connection
+    else if (method === 'eth_chainId') {
+      ws.send(
+        JSON.stringify({
+          id: request.id,
+          result: {
+            result: CHAIN_ID,
+          },
+          jsonrpc: '2.0',
+        })
+      );
+    }
+
+    // Handle other methods
+  });
+
+  ws.on('error', console.error);
+
+  ws.on('close', function () {
+    // remove subscriptions from cache
+  });
+});
+```
+
+### Client side
+
+```javascript
+var subscription = web3.eth.subscribe(
+  'logs',
+  {
+    address: '0x123456..',
+    topics: ['0x12345...'],
+  },
+  function (error, result) {
+    if (!error) console.log(result);
+  }
+);
+
+// unsubscribes the subscriptionsubscription.unsubscribe(function(error, success){
+if (success) {
+  console.log('Successfully unsubscribed!');
+}
+```
+
+### Mirror node polling:
+
+1. Maintain a list of all live socket connections. This can be done with a `key-value` store where the `key` is a unique identifier (some ws libraries provide such an id, but if needed it can be generated by the relay) of the socket object, and the `value` is the socket object itself.
+
+2. Maintain a list of all active subscriptions and the corresponding socket objects. It should have the following structure:
+
+```javascript
+  subscriptions: [
+    {
+      method: string, // newHeads, logs, etc
+      params: any,
+      subscribers: [{   // a list of all active subscriptions to the specific `method + params` combination
+        socketId: string,
+        subscription: string
+      }],
+      lastUpdated: BlockNumber // The latest block number at which this subscription was updated
+    }
+  ]
+
+```
+
+3. Create a scheduled job that loops through all saved subscriptions and calls the Mirror node Rest API and sends the response to all `subscribers`. It should be possible to configure the time interval. A throttling mechanism can be implemetented that limits the consecutive calls to the Mirror node, and at every interval only X subscriptions with the oldest `lastUpdated` should be polled.
+
+## Error Codes
```

Comment:
There should be an error case also for
- max subscription reached
- duplicate subscription id
- non matching subscription params e.g if no addresses are matched to any existing contracts
- subscription aborted (maybe when the relay is forced shut down - can we do this?)
- ... (any more)